README.md
# Easy Yummies

1.	What is the proposed name for your Web application?
-Easy Yummies
2.	Who is the target audience for your Web application?
-People with limited culinary skills and spare time to cook easy and delicious meals
3.	What problem is it intended to solve for the target audience?
-People with limited culinary skills and little spare time often find it is difficult to find recipes that are easy to cook and yummy as well. On one hand, as an editor, I will publish some easy recipes for various categories (e.g. pasta, dessert, salad, etc.) regularly. On the other hand, registered users are invited to publish their own recipes as well. All published recipes are rated, commented by registered users. In addition, “Cooking FAQ Fo-um” is created to be dedicatedly serve as a discussion forum about tips and questions on cooking.
4.	How will it meet the minimum project requirements?
-Recipes can be browsed without a login. If a user wants to comment ,rate and pin the recipes, a login is required. Similar operation is applied to the discussion forum as well. All public data will be accessible through RESTful services.
5.	Why is your proposed Web application unique or creative beyond simply meeting the minimum requirements?
-I will rank the recipes based on ratings and number of views. 

##MOST RECENT CHANGES
* build a recipe entity and RESTful webservice


## Build status


[![Build Status](https://travis-ci.org/infsci2560sp17/full-stack-web-jingyi-huang.svg?branch=master)](https://travis-ci.org/infsci2560sp17/full-stack-web-jingyi-huang)


## Changelog
[Changelog](CHANGELOG.md)


## Web Site

[Easy Yummies](https://web2560.herokuapp.com/)

## Key Features

* User can review recipes details with login
* User can publish recipes with login
* User can rate any recipe with login
* User can update his/her own profile
* User can review other user details with login
* A recipe can be retrieved from its ID in JSON format


## Project Details

### Landing Page

TODO : please provide a description of your landing page inluding a screen shot ![](screenshots/Recipes.png)

### Display Page

Review a particular recipe details including its rating history, and rate it  ![](screenshots/details.png)

### User Input Form

Publish Form : User can publish a recipe with login ![](screenshots/publish.png)

Update Form : User can update his/her profile ![](screenshots/edit.png)


## API

TODO : please provide a description of at least 1 API including a sample of request data and response data in both XML and JSON format.

### API Method 1

    GET recipes/{id}

#### Parameters

- **id** _(required)_ — The recipe id.


#### Response

{"id":4,"name":"Creamy Mushroom Pasta","recipeType":"Pasta","calories":"380 kcal","ingredients":["100g pasta","1/2 sliced onion","150g sliced mushroom","ham"],"description":["Put the pasta on to boil","Meanwhile fry sliced onions, mushroom and chopped ham a little","Drain the pasta and add the vegie sauce"],"timeNeed":"15 min","author":1}

#### Errors

All known errors cause the resource to return HTTP error code header together with a JSON array containing at least 'status' and 'error' keys describing the source of error.

- **404 Not Found** — The photo was not found.

#### Example

##### Request

    POST /v1/photos/123456/tags

##### Body

    tags=cute,puppy


##### JSON Response

```json
{
    "photoId": 123456,
    "tags": ["cute", "puppy"]
}
```

##### XML Response

```xml
<?xml version="1.0" encoding="UTF-8"?>
<PhotoTags>
    <photoId>123456</PhotoId>
        <tags>
            <tag>cute</tag>
            <tag>puppy</tag>
        </tags>
</PhotoTags>


## Technologies Used

TODO : List all technologies used in your project

- [Spring Boot](https://projects.spring.io/spring-boot/) - Takes an opinionated view of building production-ready Spring applications.
- [Thymleaf](http://www.thymeleaf.org/) - Thymeleaf is a modern server-side Java template engine for both web and standalone environments.
- [Maven](https://maven.apache.org/) - Apache Maven is a software project management and comprehension tool.

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/).

## [v1.6] - 2017-04-19
### Added
- add updating profile function for user
- sign up

## [v1.5] - 2017-04-18
### Added
- html page to review a particualr recipe details and rating history
- add rating function for registered user
- html pages to review user lists and user details
- add js to publish recipe

### Updated
- modify Recipe model, update ingredients and description attributes
- consolidate all html pages


## [v1.4] - 2017-03-25
### Added
- Rating Entity (Rating PK, primary key)
- User Entity
- Rating Repository
- User Repository
- Rating Service
- User Service
- Rating Controller
- User Controller
- UserActivity Controller
- main html
- upload css files to resource

## [v1.3] - 2017-02-16
### Added
- Recipe Entity
- Recipe Repository
- Recipe Service
- Recipe Controller
- recipes html

### Changed
- FullStackWebApplication.java, add some recipe examples to test


## [v1.2] - 2017-01-28
### Added
- Answered in README.md
    - What is the proposed name for your Web application?
    - Who is the target audience for your Web application?
    - What problem is it intended to solve for the target audience?
    - How will it meet the minimum project requirements?
    - Why is your proposed Web application unique or creative beyond simply meeting the minimum requirements?

## [v1.1] - 2017-01-21
### Added
- Cloned, configured, and compiled the scaffold project.
- Configured developer workspace.
- Deployed to Heroku.


[Unreleased]: https://github.com/infsci2560sp17/full-stack-web-jingyi-huang/compare/v1.6...HEAD
[v1.5]: https://github.com/infsci2560sp17/full-stack-web-jingyi-huang/compare/v1.5...v1.6
[v1.5]: https://github.com/infsci2560sp17/full-stack-web-jingyi-huang/compare/v1.4...v1.5
[v1.4]: https://github.com/infsci2560sp17/full-stack-web-jingyi-huang/compare/v1.3...v1.4
[v1.3]: https://github.com/infsci2560sp17/full-stack-web-jingyi-huang/compare/v1.2...v1.3
[v1.2]: https://github.com/infsci2560sp17/full-stack-web-jingyi-huang/compare/v1.1...v1.2
[v1.1]: https://github.com/infsci2560sp17/full-stack-web-jingyi-huang/compare/...v1.1


src/main/java/edu/infsci2560/FullStackWebApplication.java
package edu.infsci2560;

import edu.infsci2560.models.Rating;
import edu.infsci2560.models.RatingPk;
import edu.infsci2560.models.Recipe;
import edu.infsci2560.models.Recipe.RecipeType;
import edu.infsci2560.models.User;
import edu.infsci2560.repositories.RatingRepository;
import edu.infsci2560.repositories.RecipeRepository;

import edu.infsci2560.repositories.UserRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.ApplicationContext;

import java.util.Arrays;

@SpringBootApplication
public class FullStackWebApplication {

    private static final Logger log = LoggerFactory.getLogger(FullStackWebApplication.class);

    public static void main(String[] args) {
        ApplicationContext ctx = SpringApplication.run(FullStackWebApplication.class, args);

        RecipeRepository repository = ctx.getBean(RecipeRepository.class);
        UserRepository  userRepository = ctx.getBean(UserRepository.class);
        RatingRepository ratingRepository =ctx.getBean(RatingRepository.class);
        userRepository.save(new User("magician","123","magician@gmail.com","Pittsburgh"));
        userRepository.save(new User("megan","123","megan@gmail.com","New York"));
        userRepository.save(new User("jackSmith","123","smith_j@hotmail.com","Chicago"));
        Recipe recipe1 = new Recipe(1L, "Creamy Mushroom Pasta", RecipeType.Pasta,"380 kcal",
                Arrays.asList(new String[] {
                        "100g pasta",
                        "1/2 sliced onion",
                        "150g sliced mushroom",
                        "ham"
                }),
                Arrays.asList(new String[]{
                        "Put the pasta on to boil",
                        "Meanwhile fry sliced onions, mushroom and chopped ham a little",
                        "Drain the pasta and add the vegie sauce",
                }),"15 min");
        recipe1.setAuthor(1L);
        repository.save(recipe1);

        Recipe recipe2 = new Recipe(2L, "Tuna Salad", RecipeType.Salad,"400 kcal",
                Arrays.asList(new String[] {
                        "One can of tuna",
                        "lettuce/spinach",
                        "cheese",
                        "vinegar and salt"
                }),
                Arrays.asList(new String[]{
                        "Rinse the lettuce and cut or rip into small pieces",
                        "Add the tuna with the oil to the salad",
                        "Add vinegar and salt to your liking",
                        "Mix around with you hands",
                        "Add a bit of cheese you like"
                }),"8 min");

        recipe2.setAuthor(2L);
        repository.save(recipe2);
        //repository.save(new Recipe(3L, "Spicy Tomato Soup", RecipeType.Soup, "250 kcal", "20 min"));
        //repository.save(new Recipe(4L, "BBQ Style Burger", RecipeType.Meat, "600 kcal", "20 min"));

        ratingRepository.save(new Rating(new RatingPk(4L,3L),5));
        ratingRepository.save(new Rating(new RatingPk(4L,2L),3));
        ratingRepository.save(new Rating(new RatingPk(5L,1L),4));
        //ratingRepository.save(new Rating(new RatingPk(2L,3L),3));

    }


//    @Bean
//    public CommandLineRunner commandLineRunner(ApplicationContext ctx) {
//        return args -> {
//
//            System.out.println("Let's inspect the beans provided by Spring Boot:");
//
//            String[] beanNames = ctx.getBeanDefinitionNames();
//            Arrays.sort(beanNames);
//            for (String beanName : beanNames) {
//                System.out.println(beanName);
//            }
//
//        };
//    }
//    @Bean
//    public CommandLineRunner databaseDemo(CustomerRepository repository) {
//        return (args) -> {
//            // save a couple of customers
//            repository.save(new Customer("Jack", "Bauer"));
//            repository.save(new Customer("Chloe", "O'Brian"));
//            repository.save(new Customer("Kim", "Bauer"));
//            repository.save(new Customer("David", "Palmer"));
//            repository.save(new Customer("Michelle", "Dessler"));
//            repository.save(new Customer("Billy", "Bean"));
//
//            // fetch all customers
//            log.info("[Database Demo] Customers found with findAll():");
//            log.info("[Database Demo] -------------------------------");
//            for (Customer customer : repository.findAll()) {
//                log.info("[Database Demo] " + customer.toString());
//            }
//            log.info("");
//
//            // fetch an individual customer by ID
//            Customer customer = repository.findOne(1L);
//            log.info("[Database Demo] Customer found with findOne(1L):");
//            log.info("[Database Demo] --------------------------------");
//            log.info("[Database Demo] " + customer.toString());
//
//            // fetch customers by last name
//            log.info("[Database Demo] Customer found with findByLastName('Bauer'):");
//            log.info("[Database Demo] --------------------------------------------");
//            for (Customer bauer : repository.findByLastName("Bauer")) {
//                log.info("[Database Demo] " + bauer.toString());
//            }
//        };
//    }
}
